### ISSUE TITLE
Capturing Standard Output from Shell processes as they happen

### ISSUE THREAD

**Issue by Outrovurt** [clojure-emacs/cider#3347]
## Expected behavior
In my deps.edn I include the following library, an alternative to using `clojure.java.shell`:

`org.clojars.hozumi/clj-commons-exec {:mvn/version "1.2.0"}`

I do this so that I can redirect the output of calling a shell process to the standard output via Java's `System/out` `PrintStream`, which `clojure.java.shell/sh` doesn't allow.

If I start a REPL session and run the following:

```clojure
(require '[clj-commons-exec
    :refer (sh)])

@(sh ["testing.sh"] {:out System/out})
```
where testing.sh has the following code:

```bash
#!/bin/bash

nums="0 1 2 3 4 5"
for i in $nums ; do
	echo "i is $i"
	sleep 1
done
```
The purpose of this script is to simulate output being written by a script to standard out at different times, which I would like to see as it is reported rather than in one big go once the script has completed.

If I run the script in the Clojure REPL I set up above, I correctly see the following result, **with each line printed 1 second after the previous**:

```
i is 0
i is 1
i is 2
i is 3
i is 4
i is 5
```

## Actual behavior
Now if I run the above in the CIDER REPL, I don't see the same behaviour. Instead, when I hit the Enter key, **there is a pause for 5 seconds, after which the whole of the above is printed at once** instead of each line one at a time.

## Steps to reproduce the problem
Run the code I have supplied above in both the Clojure REPL and then the CIDER REPL in emacs.

### CIDER version information
```
;; CIDER 1.3.0-snapshot, nREPL 0.9.0
;; Clojure 1.11.1, Java 1.8.0_372
```

### Lein / Clojure CLI version

`Clojure CLI version 1.11.1.1273`

### Emacs version
GNU Emacs 28.2

### Operating system
Arch Linux

### JDK distribution
```
openjdk version "1.8.0_372"
OpenJDK Runtime Environment (build 1.8.0_372-b07)
OpenJDK 64-Bit Server VM (build 25.372-b07, mixed mode)
```

## The Question
Firstly, thanks a lot to all those of you who have been developing and maintaining CIDER over the years, it's an amazing piece of software and my Clojure/Script experience has improved considerably because of it.

Regarding the issue I have described, I don't believe it is a bug per se as much as it is down to my ignorance as to how the CIDER REPL works with respect to output streams. So really my question is, is there any way that I can get the same behaviour in the Clojure REPL to work in the CIDER REPL?

Thanks

**Comment by Outrovurt** [comment]
And thank you for the ultra-fast response!

That's a really neat trick which I never knew about. I followed the steps you described, in terminal 1 running the nrepl server and terminal 2 connecting to it and running my code. The result: everything works in terminal 1 as I was hoping to see, i.e. each line prints 1s after each previous line. I didn't even have to drop `--middleware "[cider.nrepl/cider-middleware]"`, however I tried that as well and I see exactly the same result, i.e. it makes no difference whether the middleware is included or not.

So I take it this means that `cider-nrepl` isn't the culprit? If so, what else can I try?

**Comment by Outrovurt** [comment]
Strangely it produces the following error:

```Execution error (StackOverflowError) at cider.nrepl.middleware.out/print-stream$fn (out.clj:142).```

with or without `(flush)`.

**Comment by Outrovurt** [comment]
Aha, now we're getting somewhere:

* works with `.flush`
* doesn't work without it, i.e. prints everything in one go at the very end

**Comment by Outrovurt** [comment]
Great thanks.

So is there any way I could wrap `System/out` with "something" to make it do this, rather than fork the code?

Otherwise do you have any reference on forking the lib using deps.edn? I don't think I've come across that feature before, unless you're talking about using a :local/root reference?

**Comment by Outrovurt** [comment]
> * cider rebinds *flush-on-newline* to false

No, it's `true` on my system, and even `binding` it doesn't work. Which just leaves:

> * clj-commons-exec needs to be fixed

I've just looked at the `clj-commons-exec` source code here:

[https://github.com/hozumi/clj-commons-exec/blob/master/src/clj_commons_exec.clj](url)

It appears that it doesn't use `println`/`prn` at all, which explains why `*flush-on-newline*` has no effect here.

It also shows 3 instances of calling `.flush` on the output stream, two of which occur presumably when the running process has ended, the other as it is running. The other one occurs, according to the comments, only if the input stream (specified by the `:in` option), is `System/in`, and the `:flush-input?` option is true. Otherwise, no flushing occurs:

```clojure
(defrecord FlushStreamPumper [^InputStream is ^OutputStream os]
  Runnable
  (run [_]
    (loop []
      (let [b (.read is)]
        (if (< 0 b)
          (do (.write os b)
              (.flush os)
              (recur))
          (try (.close os)
               (catch IOException _)))))))
```
In other words, you are absolutely right, the problem is entirely to do with `clj-commons-exec` and the only way to fix it is to, well, fix it.

Thanks a lot for all your help, I really appreciate it.

**Comment by Outrovurt** [comment]
Wait a second: it just struck me there's still something lacking in this explanation. The following is a summary of what we think we know so far

1. As per my original post,  `clj-commons-exec` works *outside* of CIDER, i.e. via a terminal-launched clj-REPL as expected, but not within a CIDER REPL. 

2. We then saw from [this test](https://github.com/clojure-emacs/cider/issues/3347#issuecomment-1564336705) that `cider-nrepl` "probably" (according to your subsequent comment) isn't the culprit.

3. In my last comment above I showed that neither `println` nor `prn` or anything which references `*flush-on-newline*` was being called from within `clj-common-exec`.

So it appears there is still an issue we haven't managed to uncover, specifically: why does it work in an ordinary Clojure REPL, via cider-nrepl, but not in my emacs CIDER? Could this point to `cider.el` as a possible culprit as you had alluded to earlier?

**Comment by Outrovurt** [comment]
> I don't have a favorite course of action. We could research CIDER side, just like work/research could be put into making clj-commons-exec presumably more clojure-like.
> 
> I may put time into this however no promises are made atm!
> 
> Hope you find this a reasonable conclusion for now.

If you don't know off the top of your head what it might be, I wouldn't spend any more than you already have as the library does work outside of CIDER so I can run a few things from the command-line REPL instead and that's a good interim solution. It's just much better to kick things off from CIDER for obvious reasons. I'm going to hunt around the net for clues and have posted this on the clojureverse.org form to see if anyone there has any ideas.

Thanks again for all your help.

Best, Ali

**Comment by vemv** [comment]
Feel free to re-take the conversation anytime.

Cheers - V
